File: go/pkg/manifest/session_finalization_test.go
package manifest

import (
	"encoding/json"
	"os"
	"testing"
)

func TestSessionFinalizationContract(t *testing.T) {
	data, err := os.ReadFile("../../../tests/session_finalization_vectors.json")
	if err != nil {
		t.Fatal(err)
	}
	var cases []struct {
		Name     string `json:"name"`
		Manifest string `json:"manifest"`
		Expected *bool  `json:"expected"`
	}
	if err := json.Unmarshal(data, &cases); err != nil {
		t.Fatal(err)
	}
	for _, fixture := range cases {
		t.Run(fixture.Name, func(t *testing.T) {
			parsed, err := LoadManifest(writeTmp(t, fixture.Manifest))
			if fixture.Expected == nil {
				if err == nil {
					t.Fatal("non-boolean finalizer was accepted")
				}
			} else {
				if err != nil {
					t.Fatal(err)
				}
				if parsed.Session.Commands["finish"].Finalize != *fixture.Expected {
					t.Fatal("finalizer authority was not preserved")
				}
			}
		})
	}
}

File: js/test/session_finalization.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { mkdtempSync, readFileSync, writeFileSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { loadManifest } from '../src/manifest.js';

const cases = JSON.parse(readFileSync(new URL('../../tests/session_finalization_vectors.json', import.meta.url)));
for (const fixture of cases) {
  test(`session finalization: ${fixture.name}`, () => {
    const directory = mkdtempSync(join(tmpdir(), 'toolclad-finalize-'));
    try {
      const path = join(directory, 'terminal.clad.toml');
      writeFileSync(path, fixture.manifest);
      if (fixture.expected === null) assert.throws(() => loadManifest(path));
      else assert.equal(loadManifest(path).session.commands.finish.finalize, fixture.expected);
    } finally {
      rmSync(directory, { recursive: true });
    }
  });
}
